File: mini-cursor/minicursor/providers/anthropic_backend.py
"""Native Anthropic Messages API backend (streaming, adaptive thinking, caching)."""

from __future__ import annotations

import json

import anthropic

from ..config import Config, ProviderConfig, default_context_window
from ..ui import UI
from .base import Backend, ToolCall, ToolResult, TurnResult


def _format_usage(usage) -> str:
    parts = [f"{usage.input_tokens} in", f"{usage.output_tokens} out"]
    cached = getattr(usage, "cache_read_input_tokens", 0) or 0
    if cached:
        parts.append(f"{cached} cached")
    return " / ".join(parts)


def _total_context_tokens(usage) -> int:
    """Best estimate of how many tokens the *whole* sent context used up,
    combining freshly-processed, cache-write and cache-read tokens."""
    return (
        (getattr(usage, "input_tokens", 0) or 0)
        + (getattr(usage, "cache_read_input_tokens", 0) or 0)
        + (getattr(usage, "cache_creation_input_tokens", 0) or 0)
    )


class AnthropicBackend(Backend):
    def __init__(self, pcfg: ProviderConfig, config: Config, client: anthropic.Anthropic | None = None):
        super().__init__(pcfg.model, context_window=default_context_window(pcfg))
        self.name = pcfg.name
        self.config = config  # effort/max_tokens are read live so /effort applies
        kwargs = {}
        if pcfg.base_url:
            kwargs["base_url"] = pcfg.base_url
        self.client = client or anthropic.Anthropic(**kwargs)

    def add_user_message(self, messages: list, text: str) -> None:
        messages.append({"role": "user", "content": text})

    def add_tool_results(self, messages: list, results: list[ToolResult]) -> None:
        messages.append(
            {
                "role": "user",
                "content": [
                    {
                        "type": "tool_result",
                        "tool_use_id": r.call_id,
                        "content": r.output,
                        "is_error": r.is_error,
                    }
                    for r in results
                ],
            }
        )

    def stream_turn(self, messages: list, system: str, tools: list[dict], ui: UI) -> TurnResult:
        while True:
            with self.client.messages.stream(
                model=self.model,
                max_tokens=self.config.max_tokens,
                system=[
                    {
                        "type": "text",
                        "text": system,
                        "cache_control": {"type": "ephemeral"},
                    }
                ],
                thinking={"type": "adaptive"},
                output_config={"effort": self.config.effort},
                tools=tools,
                messages=messages,
            ) as stream:
                for event in stream:
                    if event.type == "content_block_start":
                        if event.content_block.type == "thinking":
                            ui.thinking_started()
                    elif event.type == "content_block_delta":
                        if event.delta.type == "text_delta":
                            ui.stream_text(event.delta.text)
                ui.flush_stream()
                response = stream.get_final_message()

            if response.usage is not None:
                self.last_context_tokens = _total_context_tokens(response.usage)

            if response.stop_reason == "refusal":
                # Discard the (empty or partial) refused output; keep prior history.
                return TurnResult("refusal")

            # Keep full content (incl. thinking/tool_use blocks) in history.
            messages.append({"role": "assistant", "content": response.content})

            if response.stop_reason == "pause_turn":
                continue  # server-side pause; re-send to resume

            calls = [
                ToolCall(id=b.id, name=b.name, input=b.input)
                for b in response.content
                if b.type == "tool_use"
            ]
            usage = _format_usage(response.usage)
            if calls:
                return TurnResult("tool_use", calls, usage)
            if response.stop_reason == "max_tokens":
                return TurnResult("length", [], usage)
            return TurnResult("end", [], usage)

    # -- context-compaction support ---------------------------------------

    def render_round(self, round_messages: list) -> str:
        lines: list[str] = []
        for m in round_messages:
            role = m.get("role", "?")
            content = m.get("content")
            if isinstance(content, str):
                lines.append(f"{role}: {content}")
                continue
            for block in content or []:
                btype = getattr(block, "type", None)
                if btype is None and isinstance(block, dict):
                    btype = block.get("type")
                if btype == "text":
                    text = getattr(block, "text", None)
                    if text is None and isinstance(block, dict):
                        text = block.get("text", "")
                    lines.append(f"{role}: {text}")
                elif btype == "tool_use":
                    name = getattr(block, "name", None)
                    if name is None and isinstance(block, dict):
                        name = block.get("name")
                    inp = getattr(block, "input", None)
                    if inp is None and isinstance(block, dict):
                        inp = block.get("input")
                    lines.append(f"{role} called tool {name}({json.dumps(inp, default=str, ensure_ascii=False)[:500]})")
                elif btype == "tool_result":
                    out = block.get("content") if isinstance(block, dict) else getattr(block, "content", "")
                    text = out if isinstance(out, str) else json.dumps(out, default=str, ensure_ascii=False)
                    lines.append(f"tool result: {text[:1000]}")
                # "thinking" blocks are internal reasoning — skip them in the summary transcript.
        return "\n".join(lines)

    def complete_text(self, system: str, user_text: str) -> str:
        response = self.client.messages.create(
            model=self.model,
            max_tokens=2000,
            system=system,
            messages=[{"role": "user", "content": user_text}],
        )
        return "".join(b.text for b in response.content if getattr(b, "type", None) == "text")

File: tythan-code/src/core/providers/anthropicBackend.ts
/**
 * Native Anthropic Messages API backend (streaming, adaptive thinking,
 * prompt caching). Ported from the Python mini-cursor CLI's
 * anthropic_backend.py.
 */

import Anthropic from "@anthropic-ai/sdk";

import type { EffortLevel, ProviderConfig } from "../config";
import { defaultContextWindow } from "../config";
import type { NativeMessage } from "../compaction";
import type { ToolDefinition } from "../tools";
import { truncate } from "../tools";
import { Backend, makeTurnResult } from "./types";
import type { CompleteTextOptions, StreamSink, ToolCallRequest, ToolResultInput, TurnResult } from "./types";

function formatUsage(usage: Anthropic.Usage): string {
  const parts = [`${usage.input_tokens} in`, `${usage.output_tokens} out`];
  const cached = usage.cache_read_input_tokens ?? 0;
  if (cached) {
    parts.push(`${cached} cached`);
  }
  return parts.join(" / ");
}

/** Best estimate of how many tokens the *whole* sent context used up,
 * combining freshly-processed, cache-write and cache-read tokens. */
export function totalContextTokens(usage: Anthropic.Usage): number {
  return (usage.input_tokens ?? 0) + (usage.cache_read_input_tokens ?? 0) + (usage.cache_creation_input_tokens ?? 0);
}

function toAnthropicTools(tools: ToolDefinition[]): Anthropic.Tool[] {
  return tools.map((t) => ({
    name: t.name,
    description: t.description,
    input_schema: t.inputSchema as Anthropic.Tool.InputSchema,
  }));
}

export interface AnthropicAgentSettings {
  effort: EffortLevel;
  maxTokens: number;
}

export class AnthropicBackend extends Backend {
  readonly name: string;
  private readonly client: Anthropic;
  private readonly settings: AnthropicAgentSettings;

  constructor(pcfg: ProviderConfig, settings: AnthropicAgentSettings, client?: Anthropic) {
    super(pcfg.model, defaultContextWindow(pcfg));
    this.name = pcfg.name;
    this.settings = settings;
    this.client = client ?? new Anthropic({ apiKey: pcfg.apiKey, baseURL: pcfg.baseUrl });
  }

  addUserMessage(messages: NativeMessage[], text: string): void {
    messages.push({ role: "user", content: text });
  }

  addToolResults(messages: NativeMessage[], results: ToolResultInput[]): void {
    messages.push({
      role: "user",
      content: results.map((r) => ({
        type: "tool_result",
        tool_use_id: r.callId,
        content: r.output,
        is_error: r.isError,
      })),
    });
  }

  async streamTurn(
    messages: NativeMessage[],
    system: string,
    tools: ToolDefinition[],
    sink: StreamSink,
    signal?: AbortSignal,
  ): Promise<TurnResult> {
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const stream = this.client.messages.stream(
        {
          model: this.model,
          max_tokens: this.settings.maxTokens,
          system: [{ type: "text", text: system, cache_control: { type: "ephemeral" } }],
          thinking: { type: "adaptive" },
          output_config: { effort: this.settings.effort },
          tools: toAnthropicTools(tools),
          messages: messages as unknown as Anthropic.MessageParam[],
        },
        { signal },
      );

      for await (const event of stream) {
        if (event.type === "content_block_start" && event.content_block.type === "thinking") {
          sink.thinkingStarted();
        } else if (event.type === "content_block_delta" && event.delta.type === "text_delta") {
          sink.streamText(event.delta.text);
        }
      }
      sink.flushStream();
      const response = await stream.finalMessage();

      if (response.usage) {
        this.lastContextTokens = totalContextTokens(response.usage);
      }

      if (response.stop_reason === "refusal") {
        // Discard the (empty or partial) refused output; keep prior history.
        return makeTurnResult("refusal");
      }

      // Keep full content (incl. thinking/tool_use blocks) in history.
      messages.push({ role: "assistant", content: response.content });

      if (response.stop_reason === "pause_turn") {
        continue; // server-side pause; re-send to resume
      }

      const calls: ToolCallRequest[] = response.content
        .filter((b): b is Anthropic.ToolUseBlock => b.type === "tool_use")
        .map((b) => ({ id: b.id, name: b.name, input: b.input as Record<string, unknown> }));
      const usage = formatUsage(response.usage);
      if (calls.length > 0) {
        return makeTurnResult("tool_use", calls, usage);
      }
      if (response.stop_reason === "max_tokens") {
        return makeTurnResult("length", [], usage);
      }
      return makeTurnResult("end", [], usage);
    }
  }

  // -- context-compaction support ---------------------------------------

  override renderRound(roundMessages: NativeMessage[]): string {
    const lines: string[] = [];
    for (const m of roundMessages) {
      const role = m.role ?? "?";
      const content = m.content;
      if (typeof content === "string") {
        lines.push(`${role}: ${content}`);
        continue;
      }
      for (const block of (content as Array<Record<string, unknown>>) ?? []) {
        const btype = block?.type;
        if (btype === "text") {
          lines.push(`${role}: ${String(block.text ?? "")}`);
        } else if (btype === "tool_use") {
          const args = truncate(JSON.stringify(block.input ?? {}), 500);
          lines.push(`${role} called tool ${String(block.name)}(${args})`);
        } else if (btype === "tool_result") {
          const out = block.content;
          const text = typeof out === "string" ? out : JSON.stringify(out);
          lines.push(`tool result: ${truncate(text, 1000)}`);
        }
        // "thinking" blocks are internal reasoning — skip them in the summary transcript.
      }
    }
    return lines.join("\n");
  }

  override async completeText(system: string, userText: string, options?: CompleteTextOptions): Promise<string> {
    const response = await this.client.messages.create(
      {
        model: this.model,
        max_tokens: options?.maxTokens ?? 2000,
        system,
        messages: [{ role: "user", content: userText }],
      },
      { signal: options?.signal },
    );
    return response.content
      .filter((b): b is Anthropic.TextBlock => b.type === "text")
      .map((b) => b.text)
      .join("");
  }
}
